src/server/api.js

```diff
@@ -0,0 +1,372 @@
+    timestampIntervals.push({
+      start: getStartTimestampForDay(dayIndex, meta),
+      end: getEndTimestampForDay(dayIndex, meta)
+    });
+  }
+  return timestampIntervals;
+};
+
+const getDatesTimesAndPrices = (
+  close,
+  gmtoffset,
+  numberOfDays,
+  timestamp,
+  timestampIntervals
+) => {
+  let datesTimesAndPrices = [];
+  for (let dayIndex = 0; dayIndex < numberOfDays; dayIndex++) {
+    const intervalOfDatesTimesAndPrices = getDatesAndTimesForOneDay(
+      close,
+      dayIndex,
+      gmtoffset,
+      numberOfDays,
+      timestamp,
+      timestampIntervals
+    );
+    datesTimesAndPrices = datesTimesAndPrices.concat(
+      intervalOfDatesTimesAndPrices
+    );
+  }
+  return datesTimesAndPrices;
+};
+
+// numberOfDays much match the range used to obtain the intradayRes.
+const getIntradayStockData = (intradayRes, numberOfDays) => {
+  const intradayData = JSON.parse(intradayRes);
+  const result = intradayData.chart.result[0];
+  const {
+    indicators,
+    meta,
+    timestamp
+  } = result;
+  const { gmtoffset } = meta;
+
+  // Array of objects, one for each day.
+  // Each contains a start timestamp and end timestamp for that trading day.
+  const timestampIntervals = getTimestampIntervals(numberOfDays, meta);
+
+  const { close } = indicators.quote[0];
+  const datesTimesAndPrices = getDatesTimesAndPrices(
+    close,
+    gmtoffset,
+    numberOfDays,
+    timestamp,
+    timestampIntervals
+  );
+  return datesTimesAndPrices;
+};
+
+const getQueryForIntradayData = (symbol, range, interval) => {
+  return `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}` +
+  `?range=${range}&includePrePost=true&interval=${interval}` +
+  '&corsDomain=finance.yahoo.com&.tsrc=finance';
+};
+
+const requestMaxStockData = (symbol, callback) => {
+  historical({
+    // gets all data because we didn't specify from/to
+    symbol,
+    period: 'd'
+  }).then(
+    dailyData => {
+      console.log('max');
+      if (!dailyData[0]) {
+        throw new Error('Historical data was not found.');
+      }
+      callback(null, getDatesAndPrices(dailyData));
+    });
+};
+
+const requestOneDayStockData = (symbol, callback) => {
+  const queryOneDay = getQueryForIntradayData(
+    symbol,
+    QUERY_RANGE_ONE_DAY,
+    QUERY_INTERVAL_ONE_DAY
+  );
+  rp(queryOneDay)
+    .then(oneDayRes => {
+      console.log('one');
+      callback(null, getIntradayStockData(oneDayRes, ONE_DAY));
+    });
+};
+
+const requestFiveDayStockData = (symbol, callback) => {
+  const queryFiveDay = getQueryForIntradayData(symbol,
+    QUERY_RANGE_FIVE_DAY,
+    QUERY_INTERVAL_FIVE_DAY
+  );
+  rp(queryFiveDay)
+    .then(fiveDayRes => {
+      console.log('five');
+      callback(null, getIntradayStockData(fiveDayRes, FIVE_DAYS));
+    });
+};
+
+router.get('/stocks/:symbol', (req, res) => {
+  const modules = [
+    'summaryDetail',
+    'defaultKeyStatistics',
+    'financialData',
+    'price'
+  ];
+  const symbol = req.params.symbol;
+  quote({
```

Comment:
Why hasn't this call to `quote` been moved into a function that we call asynchronously?  Something like
```javascript
const requestQuote = (symbol, callback) => {
  const modules = [
    'summaryDetail',
    'defaultKeyStatistics',
    'financialData',
    'price'
  ];
  const symbol = req.params.symbol;
  quote({
    symbol,
    modules
  }).then(
    stockQuote => {
      modules.forEach(module => {
        if (!stockQuote[module]) {
          throw new Error(`Module '${module}' was not found.`);
        }
      });
      callback(null, createStock(stockQuote));
  });
}
router.get('/stocks/:symbol', (req, res) => {
  async.parallel({
    stock: callback => {
      requestQuote(symbol, callback);
    }
    maxStockData: callback => {
      requestMaxStockData(symbol, callback);
    },
    oneDayStockData: callback => {
      requestOneDayStockData(symbol, callback);
    },
    fiveDayStockData: callback => {
      requestFiveDayStockData(symbol, callback);
    }
  }, (err, results) => {
    const stock = results.stock
    for (const key in results) {
      if (!results.hasOwnProperty(key) || key === 'stock') {
        continue;
      }
      stock[key] = results[key];
    }
    res.send(stock);
  });
})
.catch(() =>
 res.status(404).send(ERROR_MESSAGE_STOCK_NOT_FOUND)
);
```
Sorry if I didn't get this perfect.  I just typed it up in GitHub and it is not tested